### PR TITLE
feat: Support mutually recursive data definitions

### PR DESCRIPTION
**Pull request by phated**
In the spirit of #711 - I've begun splitting out the changes from the "misc" section of #680

This pulls the compiler change commit directly from that PR and adds tests for the feature.

@ospencer Oh god of `wat`, can you check over the snapshots?

### CODE REVIEW

File: compiler/src/typed/typemod.re [hunk 1]
```diff
@@ -210,6 +234,16 @@ let rec map_rec_type_with_row_types = (~rec_flag, fn, decls, rem) =>
     map_rec_type(~rec_flag, fn, decls, rem)
   };
 
+let rec map2_rec_type_with_row_types = (~rec_flag, fn, decls, others, rem) =>
+  switch (decls) {
+  | [] => rem
+  | [d1, ...dl] =>
+    /*if Btype.is_row_name (Ident.name d1.typ_id) then
+        fn Trec_not d1 :: map_rec_type_with_row_types ~rec_flag fn dl rem
+      else*/
```
Comment (ospencer) [inline compiler/src/typed/typemod.re:243]:
```suggestion
```

File: compiler/src/typed/typemod.re [hunk 2]
```diff
@@ -453,12 +487,18 @@ let type_module = (~toplevel=false, funct_body, anchor, env, sstr /*scope*/) =>
     );
   };
 
-  let process_datas = (env, e, datas, attributes, loc) => {
-    let (decls, newenv) = Typedecl.transl_data_decl(env, Recursive, datas);
+  let process_datas = (env, data_decls, attributes, loc) => {
+    // e, datas
```
Comment (ospencer) [inline compiler/src/typed/typemod.re:491]:
```suggestion
```

**Review by ospencer — COMMENTED** [review]
Snapshots look good 👍

**Review by peblair — APPROVED** [review]
Thanks!

**Review by ospencer — APPROVED** [review]

**Pull request merged**